pystac/asset.py

```diff
@@ -0,0 +1,165 @@
+from copy import copy
+from typing import Any, Dict, List, Optional, TYPE_CHECKING, Union
+
+import pystac as ps
+from pystac.utils import is_absolute_href, make_absolute_href
+
+if TYPE_CHECKING:
+    from pystac.collection import Collection as Collection_Type
+    from pystac.item import Item as Item_Type
+
+
+class Asset:
+    """An object that contains a link to data associated with an Item or Collection that
```

Comment:
I haven't seen this pattern much, and I'm not sure it's a good one - this places attributes at the class level that evaluate to None, breaking their type declarations. I'd prefer not to use it, though if others think we should go that route and if it's shown as a common pattern I'd be open to it.

Reply:
I've only seen this pattern in the context of Pydantic models, which is not the case here. In the case of Pydantic, they end up doing some magic so that they don't actually end up as true class attributes in the end. I don't think we want to add them as class attributes in this PR.

Reply:
https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations

Reply:
Thanks for posting this @schwehr. I had never seen this before (probably just because I wasn't paying attention 😄 ). Now that I understand this better I would be in favor of adding these.

 It sounds like the latest version of `sphinx.autodoc` also supports docstrings on these type annotations. In that case we might could document them with their own docstrings rather than  in the `Attributes` section of the class-level docstring, which seems easier to maintain.

Reply:
I also hadn't seen this. Also seems like I was wrong about how that will create uninitialized variables:

```python
>>> class Foo:
...   x: str
...   def __init__(self):
...     pass
...
>>> f = Foo()
>>> f.x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Foo' object has no attribute 'x'
>>> Foo.x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Foo' has no attribute 'x'
```

Thanks for pointing out the PEP, makes sense to do this. I'll write an issue for a follow up PR to this to add them.

Reply:
Tracked in #321